# Examining Racial Discrimination in the US Job Market

### Background
Racial discrimination continues to be pervasive in cultures throughout the world. Researchers examined the level of racial discrimination in the United States labor market by randomly assigning identical résumés to black-sounding or white-sounding names and observing the impact on requests for interviews from employers.

### Data
In the dataset provided, each row represents a resume. The 'race' column has two values, 'b' and 'w', indicating black-sounding and white-sounding. The column 'call' has two values, 1 and 0, indicating whether the resume received a call from employers or not.

Note that the 'b' and 'w' values in race are assigned randomly to the resumes when presented to the employer.

### Exercises
You will perform a statistical analysis to establish whether race has a significant impact on the rate of callbacks for resumes.

Answer the following questions **in this notebook below and submit to your Github account**. 

   1. What test is appropriate for this problem? Does CLT apply?
   2. What are the null and alternate hypotheses?
   3. Compute margin of error, confidence interval, and p-value. Try using both the bootstrapping and the frequentist statistical approaches.
   4. Write a story describing the statistical significance in the context or the original problem.
   5. Does your analysis mean that race/name is the most important factor in callback success? Why or why not? If not, how would you amend your analysis?

You can include written notes in notebook cells using Markdown: 
   - In the control panel at the top, choose Cell > Cell Type > Markdown
   - Markdown syntax: http://nestacms.com/docs/creating-content/markdown-cheat-sheet

#### Resources
+ Experiment information and data source: http://www.povertyactionlab.org/evaluation/discrimination-job-market-united-states
+ Scipy statistical methods: http://docs.scipy.org/doc/scipy/reference/stats.html 
+ Markdown syntax: http://nestacms.com/docs/creating-content/markdown-cheat-sheet
+ Formulas for the Bernoulli distribution: https://en.wikipedia.org/wiki/Bernoulli_distribution

In [1]:
import pandas as pd
import numpy as np
from scipy import stats

In [2]:
data = pd.io.stata.read_stata('data/us_job_market_discrimination.dta')

In [3]:
# number of callbacks for white and black-sounding names
print(sum(data[data.race=='w'].call))
print(sum(data[data.race=='b'].call))

235.0
157.0


In [4]:
data.head()

id ad  education  ofjobs  yearsexp  honors  volunteer  military  empholes  \
0  b  1          4       2         6       0          0         0         1   
1  b  1          3       3         6       0          1         1         0   
2  b  1          4       1         6       0          0         0         0   
3  b  1          3       4         6       0          1         0         1   
4  b  1          3       3        22       0          0         0         0   

   occupspecific    ...      compreq  orgreq  manuf  transcom  bankreal trade  \
0             17    ...          1.0     0.0    1.0       0.0       0.0   0.0   
1            316    ...          1.0     0.0    1.0       0.0       0.0   0.0   
2             19    ...          1.0     0.0    1.0       0.0       0.0   0.0   
3            313    ...          1.0     0.0    1.0       0.0       0.0   0.0   
4            313    ...          1.0     1.0    0.0       0.0       0.0   0.0   

  busservice othservice  missind  ownership  
0        0.0        0.0      0.0             
1        0.0        0.0      0.0             
2        0.0        0.0      0.0             
3        0.0        0.0      0.0             
4        0.0        1.0      0.0  Nonprofit  

[5 rows x 65 columns]

In [5]:
data.race.unique()

array(['w', 'b'], dtype=object)

In [7]:
data.shape

(4870, 65)

<div class="span5 alert alert-success">
<p>Your answers to Q1 and Q2 here</p>
</div>

# Q1. What test is appropriate for this problem? Does CLT apply?

As the variables (race and call) are categorical hypothesis test will be used for comparing two samples.
A Two-sample T-test will be perfromed whether the means of two independent data samples differ from one another. 

In order to check whether or not CLT can be applied we shoul make sure dataset meet the following conditions.
    Sample Size: In general, a sample size of 30 is considered sufficient. In this case the sample size is 4870.
    Randomization: The data is sampled randomly.
    Independence: The sample observations must be independent of each other.
Since the assumptions and conditions are met CLT can be applied.

# Q2. What are the null and alternate hypotheses?

In a two-sample test, the null hypothesis is that the means of both groups are the same. 

H0: There is no difference between callbacks for both gorups (White and Black sounding names) 
H1: There is a diffrence between callbacks for both groups.

In [8]:
w = data[data.race=='w']
b = data[data.race=='b']
print(w.shape[0])
print(b.shape[0])

2435
2435


# Question 3 
Compute margin of error, confidence interval, and p-value. Try using both the bootstrapping and the frequentist statistical approaches.

# Bootstrapping
The idea of bootstrapping is to take many samples with replacement from the observed data set to generate a bootstrap population. Then we can use the bootstrapped population to create a sampling distribution.

In [9]:
# For White-Sounding Group, Draw some random sample with replacement and append mean to mean_temp.
w_call, sims = [], 100000
for i in range(sims):
    call_sample = np.random.choice(w['call'], replace=True, size=len(w['call']))
    sample_mean = call_sample.mean()
    w_call.append(sample_mean)
    
# Calculate bootstrapped mean and 95% confidence interval.
boot_mean = np.mean(w_call)
boot_95_ci = np.percentile(w_call, [2.5, 97.5])
print("Bootstrapped White-Sonding Group Mean Callbacks = {}, 95% CI = {}".format(boot_mean, boot_95_ci))


Bootstrapped White-Sonding Group Mean Callbacks = 0.09650836139917374, 95% CI = [0.08501027 0.10841889]


In a group of White-sounding names, with 95% confidence,
    White-sounding names get called back 8.5 - 10.8% of the time.

In [10]:
# For Black-Sounding Group, Draw some random sample with replacement and append mean to mean_temp.
b_call, sims = [], 100000
for i in range(sims):
    call_sample = np.random.choice(b['call'], replace=True, size=len(b['call']))
    sample_mean = call_sample.mean()
    b_call.append(sample_mean)
    
# Calculate bootstrapped mean and 95% confidence interval.
boot_mean = np.mean(b_call)
boot_95_ci = np.percentile(b_call, [2.5, 97.5])

print("Bootstrapped Black-Sonding Group Mean Callbacks = {}, 95% CI = {}".format(boot_mean, boot_95_ci))


Bootstrapped Black-Sonding Group Mean Callbacks = 0.06449078768491745, 95% CI = [0.0550308  0.07433265]


In a group of Black-sounding names, with 95% confidence,
    Black-sounding names get called back 5.5 - 7.4% of the time.

In [11]:
# Calculating p value
dif_bootstrap_means = [(np.mean(w_call[i])-np.mean(b_call[i])) for i in range(len(w.call))]
obs_difs = np.mean(w_call)- np.mean(b_call)
print('observed difference in means: {}'.format(obs_difs))

p = np.sum(dif_bootstrap_means >= obs_difs) / len(w_call)
print('p-value =', p)

observed difference in means: 0.03201757371425629
p-value = 0.01243


Since p-value is less than 0.05 there is a significant difference between callbacks for white and black sounding names. Thus, we reject the null hypothesis and take the alternative hypothesis.

# the frequentist statistical approaches.

In [12]:
white = sum(data[data.race=='w'].call) / len(data) 
print('Ratio of white sounding names who are called', white)
black = sum(data[data.race=='b'].call) / len(data) 
print('Ratio of black sounding names who are called', black)
w_and_b = sum(data.call) / len(data)              
print('Ratio of candidates who are called', w_and_b)

Ratio of white sounding names who are called 0.048254620123203286
Ratio of black sounding names who are called 0.03223819301848049
Ratio of candidates who are called 0.08049281314168377


In [13]:
## White-sounding names callbacks
print(white/w_and_b)

## Black-sounding names callbacks
print(black/w_and_b)

## Difference between white-sounding names and
## black-sounding names callback rate.
print(white/w_and_b - black/w_and_b)

0.5994897959183674
0.4005102040816327
0.1989795918367347


In [14]:
# Run independent t-test
ind_t_test=stats.ttest_ind(w['call'],b['call'])
print('The results of the independent t-test are: \n\tt-value = {:4.3f}\n\tp-value = {:4.3f}'.format(ind_t_test[0],ind_t_test[1]))


The results of the independent t-test are: 
	t-value = 4.115
	p-value = 0.000


According to the t-test, there is a significant difference between callbacks for white and black sounding names. Thus, we reject the null hypothesis and take the alternative hypothesis. (99% confidence)

In [15]:
np.mean(w.call)
print("The mean for the callback of white-soundng names ", np.mean(w.call))
np.mean(b.call)
print("The mean for the callback of black-soundng names ", np.mean(b.call))
b_var = b.call.var()
print("The variance for the callback of black-soundng names ", b_var)
w_var = w.call.var()
print("The variance for the callback of white-soundng names ", w_var)
avg_var = (b_var + w_var)/2
print("The avarage variance for the callbacks ", avg_var)


The mean for the callback of white-soundng names  0.09650924056768417
The mean for the callback of black-soundng names  0.0644763857126236
The variance for the callback of black-soundng names  0.060343656688928604
The variance for the callback of white-soundng names  0.08722896873950958
The avarage variance for the callbacks  0.0737863127142191


In [16]:
#Calculating confidence interval for white-sounding names

w_SE= np.std(w.call)/np.sqrt(len(w.call)) # Standard Error
ci=1.96
w_ci_low=np.mean(w.call) - ci * w_SE
w_ci_high=np.mean(w.call) + ci * w_SE
print('Confidence interval for White-soundngnames:', w_ci_low, '-', w_ci_high)

b_SE= np.std(b.call)/np.sqrt(len(b.call)) # Standard Error
b_ci_low=np.mean(b.call) - ci * b_SE
b_ci_high=np.mean(b.call) + ci * b_SE
print('Confidence interval for Black-soundngnames:', b_ci_low, '-', b_ci_high)

Confidence interval for White-soundngnames: 0.08478059723925077 - 0.10823788389611758
Confidence interval for Black-soundngnames: 0.054721252327818146 - 0.07423151909742905


# Question 4 & 5

Since p-value of the 2-sample t test < 0.05 and the confidence interval does not include 0, we con-clude that we reject the null hypothesis and there a difference between the means of number of callback of 2 samples. However, there are also many factors we have not considered in this case and there-fore the sounding of names might not be the only factor contributing to the different number of callback of 2 different races.

In [17]:
from scipy import stats
from math import sqrt
from scipy.stats import t

# Run independent t-test
ind_t_test=stats.ttest_ind(w['call'],b['call'])

# Calculate the mean difference and 95% confidence interval
N1 = sum(w['call'])
N2 = sum(b['call'])
df = (N1 + N2 - 2) # Degree of Freedom
std1 = w['call'].std()
std2 = b['call'].std()
std_N1N2 = sqrt( ((N1 - 1)*(std1)**2 + (N2 - 1)*(std2)**2) / df) # Average standard deviations between groups

diff_mean = w['call'].mean() - b['call'].mean()
MoE = t.ppf(0.975, df) * std_N1N2 * sqrt(1/N1 + 1/N2)

print('The results of the independent t-test are: \n\tt-value = {:4.3f}\n\tp-value = {:4.3f}'.format(ind_t_test[0],ind_t_test[1]))

print('Margin of Error:', MoE)
print('White-sounding Group Mean:', w['call'].mean())
print('Black-sounding Group Mean:', b['call'].mean())
print ('\nThe difference between groups is {:3.3f} [{:3.3f} to {:3.3f}] (mean [95% CI])'.format(diff_mean, diff_mean - MoE, diff_mean + MoE))


The results of the independent t-test are: 
	t-value = 4.115
	p-value = 0.000
Margin of Error: 0.05604237444040884
White-sounding Group Mean: 0.09650924056768417
Black-sounding Group Mean: 0.0644763857126236

The difference between groups is 0.032 [-0.024 to 0.088] (mean [95% CI])
